lib/agps/agps.c

```diff
@@ -0,0 +1,392 @@
+{
+	int err;
+
+	/* At this point, GPS driver or app-provided socket is assumed. */
+	if (gps_dev) {
+		return gps_agps_write(gps_dev, type_socket2gps(type), data,
+				      data_len);
+	}
+
+	err = nrf_sendto(gnss_fd, data, data_len, 0, &type, sizeof(type));
+	if (err < 0) {
+		LOG_ERR("Failed to send AGPS data to modem, errno: %d", errno);
+		err = -errno;
+	} else {
+		err = 0;
+	}
+
+	LOG_DBG("A-GPS data sent to modem");
+
+	return err;
+}
+
+static int inject_agps_type(void *agps,
+			    size_t agps_size,
+			    nrf_gnss_agps_data_type_t type,
+			    void *user_data)
+{
+	ARG_UNUSED(user_data);
+	int err;
+
+	err = send_to_modem(agps, agps_size, type);
+	if (err) {
+		LOG_ERR("Failed to send AGNSS data, type: %d (err: %d)",
+			type, err);
+		return err;
+	}
+
+	LOG_DBG("Injected AGPS data, type: %d, size: %d", type, agps_size);
+
+	return 0;
+}
+
+static int open_supl_socket(void)
+{
+	int err;
+	int proto;
+	int gai_cnt = 0;
+	uint16_t port;
+	struct addrinfo *addr;
+	struct addrinfo *info;
+
+	proto = IPPROTO_TCP;
+	port = htons(CONFIG_AGPS_SUPL_PORT);
+
+	struct addrinfo hints = {
+		.ai_family = AF_INET,
+		.ai_socktype = SOCK_STREAM,
+		.ai_protocol = proto,
+		/* Either a valid,
+		 * NULL-terminated access point name or NULL.
+		 */
+		.ai_canonname = NULL,
+	};
+
+	/* Try getaddrinfo many times, sleep a bit between retries */
+	do {
+		err = getaddrinfo(CONFIG_AGPS_SUPL_HOST_NAME, NULL, &hints,
+				  &info);
+		gai_cnt++;
+
+		if (err) {
+			if (gai_cnt < DNS_ATTEMPT_COUNT) {
+				/* Sleep between retries */
+				k_sleep(K_MSEC(1000 * gai_cnt));
+				continue;
+			}
+
+			/* Return if no success after many retries */
+			LOG_ERR("Failed to resolve IPv4 hostname %s, errno: %d",
+				CONFIG_AGPS_SUPL_HOST_NAME, err);
+			return -ECHILD;
+		}
+	} while (err);
+
+	/* Create socket */
+	supl_fd = socket(AF_INET, SOCK_STREAM, proto);
+	if (supl_fd < 0) {
+		LOG_ERR("Failed to create socket, errno %d", errno);
+		err = -errno;
+		goto cleanup;
+	}
+
+	struct timeval timeout = {
+		.tv_sec = 1,
+		.tv_usec = 0,
+	};
+
+	err = setsockopt(supl_fd,
+			 SOL_SOCKET,
+			 SO_RCVTIMEO,
+			 &timeout,
+			 sizeof(timeout));
+	if (err) {
+		LOG_ERR("Failed to setup socket timeout, errno %d", errno);
+		err = -errno;
+		goto cleanup;
+	}
+
+	/* Not connected */
+	err = -1;
+
+	for (addr = info; addr != NULL; addr = addr->ai_next) {
+		struct sockaddr *const sa = addr->ai_addr;
+
+		switch (sa->sa_family) {
+		case AF_INET6:
+			continue;
+		case AF_INET:
+			((struct sockaddr_in *)sa)->sin_port = port;
+			char ip[INET_ADDRSTRLEN];
+
+			inet_ntop(AF_INET,
+				  &((struct sockaddr_in *)sa)->sin_addr, ip,
+				  sizeof(ip));
+
+			LOG_DBG("ip %s (%x) port %d", log_strdup(ip),
+				((struct sockaddr_in *)sa)->sin_addr.s_addr,
+				ntohs(port));
+			break;
+		}
+
+		err = connect(supl_fd, sa, addr->ai_addrlen);
+		if (err) {
+			/* Try next address */
+			LOG_ERR("Unable to connect, errno %d", errno);
+			err = -errno;
```

Comment:
I thought the idea was to try to connect with the next address, and not to break the loop here? In case `connect` the final address on the list fails, `err` will be propagated to the `cleanup` section.

Reply:
Yes, of course. Missed that one. I will remove the two lines.